source/grid/manifold_lib.cc

```diff
@@ -281,71 +357,149 @@ get_new_point (const ArrayView<const Point<spacedim>> &vertices,
+  //
+  // In this step, we consider all points and directions to be embedded
+  // in a three-dimensional space.
   {
-    namespace
-    {
-      // helper function to compute a vector orthogonal to a given one.
-      template <int spacedim>
-      Point<spacedim>
-      compute_normal(const Tensor<1,spacedim> &vector)
+    Tensor<1, 3> xVec;
+    for (unsigned int c=0; c<spacedim; ++c)
+      xVec[c]=candidate[c];
+
+    // If the candidate happens to coincide with a normalized
+    // direction, we return it. Otherwise, the Hessian would be singular.
+    boost::container::small_vector<Tensor<1, 3>, 100> directions(n_points);
+    for (unsigned int i=0; i<n_points; ++i)
       {
-        AssertThrow(vector.norm() != 0.,
-                    ExcMessage("The direction parameter must not be zero!"));
-        Point<3> normal;
-        if (std::abs(vector[0]) >= std::abs(vector[1])
-            && std::abs(vector[0]) >= std::abs(vector[2]))
-          {
-            normal[1]=-1.;
-            normal[2]=-1.;
-            normal[0]=(vector[1]+vector[2])/vector[0];
-          }
-        else if (std::abs(vector[1]) >= std::abs(vector[0])
-                 && std::abs(vector[1]) >= std::abs(vector[2]))
-          {
-            normal[0]=-1.;
-            normal[2]=-1.;
-            normal[1]=(vector[0]+vector[2])/vector[1];
-          }
-        else
-          {
-            normal[0]=-1.;
-            normal[1]=-1.;
-            normal[2]=(vector[0]+vector[1])/vector[2];
-          }
-        return normal;
+        for (unsigned int c = 0; c < spacedim; ++c)
+          directions[i][c] = vertices[i][c] - center[c];
+        const double norm = directions[i].norm();
+        Assert(norm != 0.,
+               ExcMessage("One of the vertices coincides with the center. "
+                          "This is not allowed!"));
+        directions[i] /= directions[i].norm();
+        if ((xVec - directions[i]).norm_square() < tolerance*tolerance)
+          return center + rho * candidate;
       }
-    }
+
+    Tensor<1,3> vPerp;
+    Tensor<2,2> Hessian;
+    Tensor<1,2> gradient;
+    Tensor<1,2> gradlocal;
+
+    // On success we exit the loop early.
+    // Otherwise, we just take the result after max_iterations steps.
+    for (unsigned int i=0; i<max_iterations; ++i)
+      {
+        // Step 2a: Find new descent direction
+
+        // Get local basis for the estimate xVec
+        const Tensor<1,3> Clocalx = internal::compute_normal(xVec);
+        const Tensor<1,3> Clocaly = cross_product_3d(xVec, Clocalx);
+
+        // For each vertices vector, compute the tangent vector from xVec
+        // towards the vertices vector -- its length is the spherical length
+        // from xVec to the vertices vector.
+        // Then compute its contribution to the Hessian.
+        gradient = 0.;
+        Hessian = 0.;
+        for (unsigned int i=0; i<n_points; ++i)
+          if (weights[i]>tolerance)
```

Comment:
Could you please change this into `if (std::abs(weights[i]) > 1e-15)`? Besides allowing for negative weights, we should also have a stricter tolerance than `1e-10` here. I think we can be pretty close to roundoff since we either are zero or not.

Reply:
Probably better to use std::numeric_limits<double>::epsilon() * 10 or alike